README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# compstatr <img src="man/figures/logo.png" align="right" />

[![lifecycle](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![Travis-CI Build
Status](https://travis-ci.com/slu-openGIS/compstatr.svg?branch=master)](https://travis-ci.com/slu-openGIS/compstatr)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/slu-openGIS/compstatr?branch=master&svg=true)](https://ci.appveyor.com/project/chris-prener/compstatr)
[![Coverage
status](https://codecov.io/gh/slu-openGIS/compstatr/branch/master/graph/badge.svg)](https://codecov.io/github/slu-openGIS/compstatr?branch=master)
[![CRAN\_status\_badge](http://www.r-pkg.org/badges/version/compstatr)](https://cran.r-project.org/package=compstatr)
[![cran
checks](https://cranchecks.info/badges/worst/compstatr)](https://cran.r-project.org/web/checks/check_results_compstatr.html)
[![DOI](https://zenodo.org/badge/105331568.svg)](https://zenodo.org/badge/latestdoi/105331568)

The goal of `compstatr` is to provide a suite of tools for working with
crime data made public by the City of St. Louis’ [Metropolitan Police
Department](http://www.slmpd.org).

## Motivation

Among cities in the United States, St. Louis has the distinction of
having the highest or one of the highest violent crime and homicide
rates since 2010. It therefore presents an important site for social
scientists, public health researchers, and health care providers as well
as policy makers to understand the effects of violent crime on urban
communities.

The City’s crime data, however, are difficult to work with and present a
number of challenges for researchers. These data are inconsistently
organized, with all data before 2013 and some months of 2013 itself
having eighteen variables. Beginning during 2013, most (but not all)
months have twenty variables, many of which are named differently from
their pre-2014 counterparts. These inconsistencies, and the fact that
working with their data requires managing over 120 spreadsheets that
each download with with a `.html` file extension, are the motivating
force behind `compstatr`.

We therefore provide a set of tools for accessing, preparing, editing,
and mapping St. Louis [Metropolitan Police
Department](http://www.slmpd.org) (SLMPD) crime data, which are
available [on their website](http://www.slmpd.org/Crimereports.shtml) as
`.csv` files. The categorization tools that are provided will work with
any police department that uses 5 and 6 digit numeric codes to identify
specific crimes.

## What’s New in v0.2.1?

Version v0.2.1 addresses a number of bugs, and returns `compstatr` to
CRAN.

## Installation

The easiest way to get `compstatr` is to install it from CRAN:

``` r
install.packages("compstatr")
```

The development version of `compstatr` can be accessed from GitHub with
`remotes`:

``` r
# install.packages("remotes")
remotes::install_github("slu-openGIS/compstatr")
```

## Usage

We’ll start with loading the `compstatr` package:

``` r
> library(compstatr)
```

### Data Access - Read Tables Directly into R

As of version `v0.2.0`, data tables can be scraped and read directly
into `R` without manually downloading them first. They are read from the
St. Louis Metropolitan Police Department’s
[website](http://www.slmpd.org/Crimereports.shtml) and imported directly
as objects in `R`’s global environment. To identify the last available
month:

``` r
> cs_last_update()
[1] "May 2019"
```

To enable scraping, an index of the available data needs to be created.
Doing this is optional but highly recommended to improve performance:

``` r
> # create index
> i <- cs_create_index()
```

This index is used by `cs_get_data()` to find the requested table or
tables, post a request via the SLMPD website’s form system, and then
download your data:

``` r
> # download single month
> may17 <- cs_get_data(year = 2017, month = "May", index = i)
>
> # download full year
> yearList17 <- cs_get_data(year = 2017, index = i)
```

Once data are downloaded, they need to be validated and standardized
before proceeding with analysis.

### Data Access - Use Tables Downloaded Manually

While scraping is now an option, St. Louis data can still be downloaded
month-by-month from [SLMPD](http://www.slmpd.org/Crimereports.shtml).
`compstatr` assumes that only one year of crime data (or less) is
included in specific folders within your project. These next examples
assume you have downloaded all of the data for 2017 and 2018, and saved
them respectively in `data/raw/2017` and `data/raw/2018`.

The function `cs_prep_data()` can be used to rename files, which may be
downloaded with the wrong file extension (`January2018.csv.html`). Once
downloaded you can load them into what we call year-list objects:

``` r
> cs_prep_year(path = "data/raw/2017")
>
> yearList17 <- cs_load_year(path = "data/raw/2017")
```

Once data are downloaded, they need to be validated and standardized
before proceeding with analysis.

### Data Preparation

Both the data downloaded manually as well as the tables scraped from
SLMPD’s website are inconsistently organized. Problems that need to be
addressed prior to collapsing a year-list into a single object can be
identified with `cs_validate()`:

``` r
> cs_validate(yearList17, year = 2017)
[1] FALSE
```

If a `FALSE` value is returned, the `vebose = TRUE` argument provides
additional detail:

``` r
> cs_validate(yearList17, year = 2017, verbose = TRUE)
# A tibble: 12 x 8
   namedMonth codedMonth valMonth codedYear valYear oneMonth varCount valVars
   <chr>      <chr>      <lgl>        <int> <lgl>   <lgl>    <lgl>    <lgl>  
 1 January    January    TRUE          2017 TRUE    TRUE     TRUE     TRUE   
 2 February   February   TRUE          2017 TRUE    TRUE     TRUE     TRUE   
 3 March      March      TRUE          2017 TRUE    TRUE     TRUE     TRUE   
 4 April      April      TRUE          2017 TRUE    TRUE     TRUE     TRUE   
 5 May        May        TRUE          2017 TRUE    TRUE     FALSE    NA     
 6 June       June       TRUE          2017 TRUE    TRUE     TRUE     TRUE   
 7 July       July       TRUE          2017 TRUE    TRUE     TRUE     TRUE   
 8 August     August     TRUE          2017 TRUE    TRUE     TRUE     TRUE   
 9 September  September  TRUE          2017 TRUE    TRUE     TRUE     TRUE   
10 October    October    TRUE          2017 TRUE    TRUE     TRUE     TRUE   
11 November   November   TRUE          2017 TRUE    TRUE     TRUE     TRUE   
12 December   December   TRUE          2017 TRUE    TRUE     TRUE     TRUE   
```

In this case, we have the wrong number of variables for the month of May
(in this case there are 26). We can fix this by using `cs_standardize()`
to create the correct number of columns (20) and name them
appropriately:

``` r
> # standardize 
yearList17 <- cs_standardize(yearList17, month = "May", config = 26)
>
> # confirm data are now valid
> cs_validate(yearList17, year = 2017)
[1] TRUE
```

For 2013 and prior years, there will be only 18 variables. The 2013 data
need to be fixed month by month because there are some correct months,
but years 2008 through 2012 can be fixed en masse:

``` r
> yearList08 <- cs_standardize(yearList08, config = 18, month = "all")
```

Once the data have been standardized, we can collapse them into a single
object with `cs_collapse()`:

``` r
> reports17 <- cs_collapse(yearList17)
```

This gives us all of the crimes reported in 2017. However, there will be
crimes that were reported that year that occurred in prior years, and
there may also be crimes reported in 2018 that took place in our year of
interest. We can address both issues (assuming we have the next year’s
data) with `cs_combine()`:

``` r
> # load and standardize 2018 data
> cs_prep_year(path = "data/raw/2018")
> yearList18 <- cs_load_year(path = "data/raw/2018")
> cs_validate(yearList18, year = 2018)
[1] TRUE
> reports18 <- cs_collapse(yearList18)
>
> # combine 2017 and 2018 data
crimes17 <- cs_combine(type = "year", date = 2017, reports17, reports18)
```

We now have a tibble containing all of the known crimes that occurred in
2017 (including those reported in 2018).

### Data Wrangling and Mapping

Once we have the data prepared, we can easily pull out a specific set of
crimes to inspect further. For example, we could identify homicides. In
the next few examples, we’ll use the `january2018` example data that
comes with the package. We’ll start by using `cs_filter_crimes()` to
select only homicides as well as `cs_filter_count()` to remove any
unfounded incidents:

``` r
> # load dependencies
> library(compstatr)
> library(ggplot2)
> library(magrittr)
> library(mapview)
> 
> # subset homicides and removed unfounded incidents
> janHomicides <- january2018 %>%
+   cs_filter_count(var = Count) %>%
+   cs_filter_crime(var = Crime, crime = "homicide")
```

Next, we’ll check for missing spatial data with `cs_missingXY()`:

``` r
> # identify missing spatial data
> janHomicides <- cs_missingXY(janHomicides, varX = XCoord, varY = YCoord, newVar = missing)
> 
> # check for any TRUE values
> table(janHomicides$missing)
```

We don’t have any missing spatial data in this example, but if we did we
would need to remove those observations with `dplyr::filter()` (or
another subsetting tool). Finally, we can project and map our data:

``` r
> # project data
> janHomicides_sf <- cs_projectXY(janHomicides, varX = XCoord, varY = YCoord)
> 
> # preview data
> mapview(janHomicides_sf)
```

<img src="man/figures/homicide_mapview.png" width="100%" />

These data can also be mapped using `ggplot2` once they have been
projected:

``` r
> library(ggplot2)
> ggplot() + 
+   geom_sf(data = janHomicides_sf, color = "red", fill = NA, size = .5)
```

<img src="man/figures/homicide_map.png" width="40%" />

## Non St. Louis Data

If you work with data from other police departments, the `cs_crime()`,
`cs_crime_cat()`, and `cs_filter_crime()` functions may be useful for
identifying, grouping, and subsetting by crime so long as they use a
standard set of 5 and 6 digit codes based on the UCR system (e.g.
`31111` (robbery with a firearm) or `142320` (malicious destruction of
property)).

## Acknowledgements

We wish to thank Taylor Braswell for his significant efforts compiling
Stata code early in this project. Taylor’s code was used as a reference
when developing this package, and many of the functions reflect issues
that he worked to identify.

## Contributor Code of Conduct

Please note that this project is released with a [Contributor Code of
Conduct](https://github.com/slu-openGIS/compstatr/blob/master/.github/CODE_OF_CONDUCT.md).
By participating in this project you agree to abide by its terms.


NEWS.md
# compstatr 0.2.1

* address bug where categorize and filter functions failed when the `var` argument was not an integer variable
* address testing rules related to `epsg` values in new `sf` release
* ensure that upcoming changes with `dplyr` and `tibble` do not impact package

# compstatr 0.2.0

* add functionality for directly scraping crime data into `R`:
    * add `cs_last_update` - checks date of last data release by SLMPD
    * add `cs_create_index` - creates a full tibble of available month tables
    * add `cs_get_data` - downloads a month or year worth of crime tables
    * There is one breaking change for users - to enable these functions, all variable names are standardized into `snake_case` at import. This will impact code that has been written using a prior version of `compstatr`.
* update `README` and `index.Rmd` with "What's New?" sections

# compstatr 0.1.1

* **initial CRAN submission**
* add Zenodo release badge to `README`
* update installation instructions on `README` and on the `pkgdown` site
* add `cran-comments.md`
* in response to initial feedback from Matthias Sterrer (CRAN):
    * add `cs_example()` function for creating a sample year worth of `.csv` files 
    * add executable examples for `cs_prep_year()`, `cs_load_year()`, and `cs_projectXY`
    * clarify `DESCRIPTION` for package
* in response to additional feedback from Swetlana Herbrandt (CRAN):
    * added a link to the `DESCRIPTION` file to the SLMPD data source
    * ensured that the examples in `example.R` and `create.R` read from and write to a temporary directory
* add a function `cs_address()` to facilitate concatenation of street addresses prior to geocoding

# compstatr 0.1.0

* Added a `NEWS.md` file to track changes to the package.
* Initial package development including package infrastructure, functions, and sample data


R/create.R
#' Prepare Raw Data
#'
#' @description Data downloaded from the St. Louis Metropolitan Police Department
#'    are downloaded with incorrect file paths - e.g. \code{January2008.CSV.html}.
#'    This function iterates over all files in a given path and replaces their
#'    file extensions. Thus \code{January2008.CSV.html} will be replaced by
#'    \code{january2008.csv}. There should be no more than 12 files in a given path,
#'    and all should correspond to the same year.
#'
#' @usage cs_prep_year(path, verbose = FALSE)
#'
#' @param path File path where raw STLMPD data are
#' @param verbose If \code{TRUE}, returns a tibble with results; otherwise if \code{FALSE}, no
#'     output is returned.
#'
#' @return A tibble containing old file names and new file names for reference is \code{verbose = TRUE}.
#'     Otherwise, no output is returned. This function will change all problematic filenames in the
#'     specified path.
#'
#' @examples
#' # create temporary directory
#' tmpdir <- tempdir()
#' fs::dir_create(paste0(tmpdir,"/data/"))
#'
#' # load sample files into temporary directory
#' cs_example(path = paste0(tmpdir,"/data/"))
#'
#' # list files
#' list.files(paste0(tmpdir,"/data/"))
#'
#' # prep sample files
#' cs_prep_year(path = paste0(tmpdir,"/data/"))
#'
#' # list files again
#' list.files(paste0(tmpdir,"/data/"))
#'
#' # delete data
#' fs::dir_delete(paste0(tmpdir,"/data/"))
#'
#' # create temporary directory
#' fs::dir_create(paste0(tmpdir,"/data/"))
#'
#' # load sample files into temporary directory
#' cs_example(path = paste0(tmpdir,"/data/"))
#'
#' # prep sample files
#' cs_prep_year(path = paste0(tmpdir,"/data/"), verbose = TRUE)
#'
#' # delete data again
#' fs::dir_delete(paste0(tmpdir,"/data/"))
#'
#' @importFrom dplyr as_tibble
#' @importFrom dplyr filter
#' @importFrom fs file_move
#' @importFrom purrr map
#' @importFrom purrr map_chr
#' @importFrom stringr str_c
#' @importFrom stringr str_detect
#' @importFrom stringr str_replace
#'
#' @export
cs_prep_year <- function(path, verbose = FALSE){

  # check parameters
  if (is.logical(verbose) == FALSE){
    stop("The 'verbose' parameter only accepts 'TRUE' or 'FALSE' as valid arguments.")
  }

  # create vector of filenames
  files <- list.files(path)

  # check number of files
  if (length(files) > 12){

    stop('There are too many files in the specified folder. Edit crime files in yearly batches of 12 monthly files.')

  } else if (length(files) < 12){

    warning('There are fewer than 12 files in the specified folder. You are only editing a partial year.')

  }

  # detect html in file extensions
  html <- stringr::str_detect(files, pattern = ".html$")

  # create data frame of files and file extensions
  data <- data.frame(files = files, html = html, stringsAsFactors = FALSE)

  # subset data frame and convert to vector
  data <- dplyr::filter(data, html == TRUE)
  problemFiles <- as.vector(data$files)

  # iterate and produce optional output
  if (verbose == TRUE){

    # iterate over each filename, renaming it and coverting to lowercase
    problemFiles %>%
      split(problemFiles) %>%
      purrr::map_chr(~ cs_edit_filename(path = path, file = .x)) -> changes

    # create vector of new filenames
    orignal <- names(changes)
    names(changes) <- NULL

    # create output
    out <- dplyr::as_tibble(data.frame(
      original = orignal,
      new = changes,
      stringsAsFactors = FALSE
    ))

    # return output
    return(out)

  } else if (verbose == FALSE){

    # iterate over each filename, renaming it and coverting to lowercase
    problemFiles %>%
      split(problemFiles) %>%
      purrr::map(~ cs_edit_filename(path = path, file = .x)) -> out

  }

}

# edit an individual file name
cs_edit_filename <- function(path, file){

  # construct a new file name that is all lower case, removes .html from end
  newFile <- tolower(stringr::str_replace(file, pattern = ".html$", replacement = ""))

  # create new file paths, adding a forward slash between path and filename if necessary
  if (stringr::str_detect(path, pattern = "/$") == FALSE){

    filePath <- stringr::str_c(path, "/", file)
    newPath <- stringr::str_c(path, "/", newFile)

  } else {

    filePath <- stringr::str_c(path, file)
    newPath <- stringr::str_c(path, newFile)

  }

  # rename file
  fs::file_move(filePath, newPath)

  # return output
  return(newFile)

}

#' Create Year List Object
#'
#' @description \code{cs_load_year} is used to load a set of \code{.csv} files
#'    contained in the given directory. This should be used to load a full
#'    year worth of data or a partial year. There should be no more than 12
#'    files in a given path, and all should correspond to the same year. All
#'    columns will be read in as character data in order to address inconsistencies
#'    in how the data are created. When \link{cs_collapse} is executed, variables
#'    will be converted numeric when doing so is applicable.
#'
#' @usage cs_load_year(path)
#'
#' @param path A file path
#'
#' @return A year-list object containing 12 tibbles - one per month - worth
#'    of crime data stored within a list.
#'
#' @examples
#' # create temporary directory
#' tmpdir <- tempdir()
#' fs::dir_create(paste0(tmpdir,"/data/"))
#'
#' # load sample files into temporary directory
#' cs_example(path = paste0(tmpdir,"/data/"))
#'
#' # prep sample files
#' cs_prep_year(path = paste0(tmpdir,"/data/"))
#'
#' # load sample files
#' yearList17 <- cs_load_year(path = paste0(tmpdir,"/data/"))
#'
#' # delete data
#' fs::dir_delete(paste0(tmpdir,"/data/"))
#'
#' # print year-list object
#' yearList17
#'
#' @importFrom dplyr %>%
#' @importFrom purrr map
#' @importFrom readr cols
#' @importFrom readr col_character
#' @importFrom readr read_csv
#' @importFrom stringr str_sub
#'
#' @export
cs_load_year <- function(path){

  # create list of all files at path that are csv
  files <- dir(path = path, pattern = "*.csv")

  # check number of files
  if (length(files) > 12){

    stop('There are too many files in the specified folder. Load crime files in yearly batches of 12 monthly files.')

  } else if (length(files) < 12){

    warning('There are fewer than 12 files in the specified folder. You are only loading a partial year.')

  }

  # read csv files into year list objects
  files %>%
    purrr::map(~ suppressMessages(suppressWarnings(readr::read_csv(file.path(path, .), col_types = readr::cols(.default = readr::col_character()))))) -> out

  # clean-up variable names
  out %>%
    purrr::map(janitor::clean_names) -> out

  # create list of months associated with year list object items
  out %>%
    purrr::map(cs_identifyMonth) -> nameList

  # convert list of months to vector
  nameVector <- unlist(nameList, recursive = TRUE, use.names = TRUE)

  # apply vector to data
  names(out) <- nameVector

  # add new class
  class(out) <- append(class(out), "cs_year_list")

  # return year list object
  return(out)

}

# Extract Month of a Given Year List Object Item
#
# @description  This uses the value of the first observation's coded month as the basis for
#     identifying which month the data are from.
#
# @param .data A year list object name
# @param read A logical scalar; if \code{TRUE}, return output structured for initial reading of
#    data into R. If \code{FALSE}, use for data validation.
#
cs_identifyMonth <- function(.data, read = TRUE){

  # depending on number of columns, the CodedMonth variable is named differently
  # the if elseif statements pull the first value from CodedMonth

  if (length(.data) == 18){

    monthVal <- .data$month_reportedto_mshp[1]

  } else if (length(.data) == 20 | length(.data) == 26){

    monthVal <- .data$coded_month[1]

  }

  # extract the last two digits from the coded month value
  monthString <- stringr::str_sub(monthVal, start = -2)

  if (read == TRUE){

    # convert those last two digits into a string month name
    out <- cs_matchMonth(monthString)

  } else if (read == FALSE){

    out <- as.numeric(monthString)

  }

  # return output
  return(out)

}

# Match Extract Month with Month Name
#
# @description Based on the result of cs_identifyMonth, this function returns the
#     appropriate string name.
#
# @param x The last two characters of the first observation's coded month value
#
cs_matchMonth <- function(x){

  # the last two digits from two digits from the coded month value are passed to this function as x
  # depending on the value, the correct month string name is returned

  if (x == "01") {

    name <- "January"

  } else if (x == "02") {

    name <- "February"

  } else if (x == "03") {

    name <- "March"

  } else if (x == "04") {

    name <- "April"

  } else if (x == "05") {

    name <- "May"

  } else if (x == "06") {

    name <- "June"

  } else if (x == "07") {

    name <- "July"

  } else if (x == "08") {

    name <- "August"

  } else if (x == "09") {

    name <- "September"

  } else if (x == "10") {

    name <- "October"

  } else if (x == "11") {

    name <- "November"

  } else if (x == "12") {

    name <- "December"

  }

  # return output
  return(name)

}
